runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
       windowsToConsider.addAll(processElement(windowToMergeResult, value));
     }
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```

Comment:
Sorry, I didn't properly explain myself:

In the case of small windows and high skew from a source, the number of windows in a bundle is ~ number of elements in the bundle. This is not a great situation for many other reasons, but we empirically see this happening a lot for various sources.